### PR TITLE
Formation assignments entity

### PR DESCRIPTION
**Pull request by PetarTodorovv**
**Description**
Introducing a new formation assignment entity

Changes proposed in this pull request:
- new `formation_assignments` DB entity
- CRUD operations for repo and service layers 
- Formation assignments list functionality as part of the `formation` resolver(sub-resolver)
- Data loaders for the new formation assignments
- Unit tests

**Pull Request status**
- [x] Implementation
- [x] Unit tests
- [x] [N/A] Integration tests
- [x] `chart/compass/values.yaml` is updated 
- [x] Mocks are regenerated, using the automated script

### CODE REVIEW

File: components/director/internal/domain/formation/resolver.go [hunk 1]
```diff
@@ -30,6 +32,20 @@ type Converter interface {
 	MultipleToGraphQL(in []*model.Formation) []*graphql.Formation
 }
 
+// FormationAssignmentService missing godoc
```
Comment (la4ezar) [inline components/director/internal/domain/formation/resolver.go:35]: Add godoc.

File: components/director/internal/domain/formation/resolver.go [hunk 2]
```diff
@@ -30,6 +32,20 @@ type Converter interface {
...
+//go:generate mockery --name=FormationAssignmentService --output=automock --outpkg=automock --case=underscore --disable-version-string
+type FormationAssignmentService interface {
+	ListByFormationIDs(ctx context.Context, formationIDs []string, pageSize int, cursor string) ([]*model.FormationAssignmentPage, error)
+	GetForFormation(ctx context.Context, id, formationID string) (*model.FormationAssignment, error)
+}
+
+// FormationAssignmentConverter missing godoc
```
Comment (la4ezar) [inline components/director/internal/domain/formation/resolver.go:42]: Add godoc.

File: components/director/internal/domain/formation/resolver.go [hunk 3]
```diff
@@ -233,3 +253,92 @@ func (r *Resolver) UnassignFormation(ctx context.Context, objectID string, objec
...
+		fas := r.formationAssignmentConv.MultipleToGraphQL(page.Data)
+
+		gqlFormationAssignments = append(gqlFormationAssignments, &graphql.FormationAssignmentPage{Data: fas, TotalCount: page.TotalCount, PageInfo: &graphql.PageInfo{
+			StartCursor: graphql.PageCursor(page.PageInfo.StartCursor),
+			EndCursor:   graphql.PageCursor(page.PageInfo.EndCursor),
+			HasNextPage: page.PageInfo.HasNextPage,
+		}})
+	}
+
+	err = tx.Commit()
```
Comment (la4ezar) [inline components/director/internal/domain/formation/resolver.go:308]: if err := tx.Commit(); err != nil { ... }

File: components/director/internal/domain/formation/resolver.go [hunk 4]
```diff
@@ -233,3 +253,92 @@ func (r *Resolver) UnassignFormation(ctx context.Context, objectID string, objec
...
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	return gqlFormationAssignments, nil
+}
+
+// FormationAssignment missing godoc
```
Comment (la4ezar) [inline components/director/internal/domain/formation/resolver.go:316]: Add godoc.

File: components/director/internal/domain/formation/resolver.go [hunk 5]
```diff
@@ -233,3 +253,92 @@ func (r *Resolver) UnassignFormation(ctx context.Context, objectID string, objec
...
+
+	formationAssignment, err := r.formationAssignmentSvc.GetForFormation(ctx, id, obj.ID)
+	if err != nil {
+		if apperrors.IsNotFoundError(err) {
+			return nil, tx.Commit()
+		}
+		return nil, err
+	}
+
+	err = tx.Commit()
```
Comment (la4ezar) [inline components/director/internal/domain/formation/resolver.go:338]: if err := tx.Commit(); err != nil { ... }

File: components/director/internal/domain/formationassignment/service.go [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+	}
+
+	exists, err := s.repo.Exists(ctx, id, tenantID)
+	if err != nil {
+		return errors.Wrapf(err, "while ensuring formation assignment with ID: %q exists", id)
+	} else if !exists {
+		return apperrors.NewNotFoundError(resource.FormationAssignment, id)
+	}
+
+	err = s.repo.Update(ctx, in.ToModel(id, tenantID))
```
Comment (la4ezar) [inline components/director/internal/domain/formationassignment/service.go:137]: if err := ...; err != nil { ... }

File: components/director/internal/domain/formationassignment/service.go [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+// Create creates a formation assignment using `in`
+func (s *service) Create(ctx context.Context, in *model.FormationAssignmentInput) (string, error) {
+	formationAssignmentID := s.uidSvc.Generate()
+	tenantID, err := tenant.LoadFromContext(ctx)
+	if err != nil {
+		return "", errors.Wrapf(err, "while loading tenant from context")
+	}
+	log.C(ctx).Debugf("ID: %q generated for formation assignment for tenant with ID: %q", formationAssignmentID, tenantID)
+
+	log.C(ctx).Infof("Creating formation assignment with source: %q and source type: %q, and target: %q with target type: %q", in.Source, in.SourceType, in.Target, in.TargetType)
```
Comment (la4ezar) [inline components/director/internal/domain/formationassignment/service.go:55]: Maybe these types of logs will be helpful in the other functions too(Delete, Update, etc.).

**Review by la4ezar — CHANGES_REQUESTED** [review]

**Review by la4ezar — APPROVED** [review]

File: components/director/internal/domain/formationassignment/repository_test.go [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+		},
+		RepoConstructorFunc:       formationassignment.NewRepository,
+		ModelEntity:               faModel,
+		DBEntity:                  faEntity,
+		NilModelEntity:            nilFormationAssignmentModel,
+		IsGlobal:                  true,
+		DisableConverterErrorTest: true,
+	}
+
+	suite.Run(t)
```
Comment (nyordanoff) [inline components/director/internal/domain/formationassignment/repository_test.go:41]: In all test suites there is a comment like: `// This test suite is not suitable for entities with embedded tenant in them.`

This, however, is outdated and misleading. Can I ask you to clean up those comments in the test suites.

**Review by nyordanoff — CHANGES_REQUESTED** [review]

File: components/director/internal/domain/formationassignment/converter.go [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+type converter struct{}
+
+// ToGraphQL converts from internal model to GraphQL output
+func (c *converter) ToGraphQL(in *model.FormationAssignment) *graphql.FormationAssignment {
+	if in == nil {
+		return nil
+	}
+	marshalledValue, err := json.Marshal(in.Value)
+	if err != nil {
+		return nil
```
[the remark this answers is not in the document]
Comment (PetarTodorovv) [reply components/director/internal/domain/formationassignment/converter.go:25]: If we return error we'll then need to check for error and if that's the case we need to return nil. This method is used in the resolver that returns `*graphql.FormationAssignment` and in the `MultipleToGraphQL` method that also return `[]*graphql.FormationAssignment`. That's why here I directly return nil(and ignore the marshal error)

**Review by nyordanoff — APPROVED** [review]

**Pull request merged**